proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,2245 @@
+
+* ``Just``, when used as a term, is given the type ``Just :: a :p-> Maybe  a``
+* ``(:)``, when used as a term, is given the type ``(:) :: a :p-> [a]
+  :q-> [a]``
+* With ``data U a where U :: a -> U a``, when ``U`` is used as a term, it
+  is given the type ``U :: a -> U a``
+* With ``data P a b where P :: a ->. b -> U a b``, when ``P`` is used
+  as a term, it is given the type ``P :: a :p-> b -> U a b``
+
+See also `η-expansion`_ for a conceptually simpler alternative which
+turns out not to be complete. See `More multiplicities`_ for
+considerations in a more general setting.
+
+Monomorphisation
+----------------
+
+.. _Monomorphisation:
+
+We want that code which doesn't use ``-XLinearTypes`` work as it did
+before. However, since constructors are now linear by default, and
+generalised due to the rule of `Linear constructors`_, we need to
+prevent multiplicity variables to be visible to the unsuspecting user.
+
+To that effect, much like is done for levity variables, wherever type
+variables would be generalised, remaining multiplicity variables are
+monomorphised to ``ω``. This way, ``f = Just`` is inferred to have
+type ``a -> Maybe a`` as before.
+
+This also address a more serious compatibility issue. Consider the
+following Haskell98 code
+
+::
+
+   class Category arr where
+     (.) :: b `arr` c -> a `arr` b -> a `arr` c
+
+   instance Category (->) where
+     f . g = \x -> f (g x)
+
+   f = Just . Just $ 1
+
+The type checker infers that ``Just . Just`` is of type ``a :p-> Maybe
+(Maybe a)`` for some ``p`` such that ``Category (:p->)``. However,
+there is no ``Category`` instance for an arbitrary ``p`` (nor for
+``p=1`` as would be the inferred type without the generalisation rule
+of the `Linear constructors`_ section). But monomorphising to ``p=ω``,
+lets the constraint solver pick the intended ``Category`` instance.
+
+Strict & unpacked fields
+------------------------
+
+Strict fields, whether unpacked or not, are treated, for the purpose of linearity, just like
+regular fields, *e.g.*
+
+::
+
+    data S a = S !a (S a)
+
+    -- S :: a ->. S a ->. S a
+    --
+    -- Or, polymorphised when used as a term:
+    --
+    -- S :: forall p q. a :p->. S a :q-> S a
+
+::
+
+    data T a = T {-# UNPACK #-}!(a, a) a
+
+    -- T :: (a, a) ->. a ->. T a
+    --
+    -- Or, polymorphised when used as a term:
+    --
+    -- T :: forall p q. (a, a) :p->. a :q-> T a
+
+Base
+----
+
+Because linear functions only strengthen the contract of unrestricted
+functions, a number of functions of ``base`` can get a more precise
+type. However, for pedagogical reasons, to prevent linear types from
+interfering with newcomers' understanding of the ``Prelude``, this
+proposal does not modify ``base``. Instead, we expect that users will
+publish new libraries on Hackage including more precisely typed
+``base`` functions. One such library has already started `here
+<https://github.com/tweag/linear-base>`_.
```

Comment:
14. Will this alternative `base` be usable in practice? Presumably, your definition of `Monad` has changed. How does that affect `do` notation?

What I'm worried about here is that we accept this proposal with the expectation of not modifying `base`, and then it becomes clear that a separate `base` is simply unusable and so we have to rush the update of `base`.